docs/interface.md

```diff
@@ -0,0 +1,56 @@
+# The Interface
+
+All interactions with the Exercism website are handled automatically. Test runners have the single responsibility of taking a solution, running its tests and returning a standarised output.
+
+## Execution
+
+- An test runner should provide an executable script. You can find more information in the [docker.md](https://github.com/exercism/automated-mentoring-support/blob/master/docs/docker.md) file.
+- The script will receive two parameters:
+  - The slug of the exercise (e.g. `two-fer`).
+  - A path to a directory containing the submitted file(s) (with a trailing slash).
+- The script must write a file to that directory named `report.json`
+
+## Output format
+
+The `report.json` file should be structured as followed:
+
+```json
+{
+  "status": "...",
+  "tests": [
+    {
+      "name": "Test that the thing works",
```

Comment:
Are there any plans for handling empty names? 

It's not possible nor common to name individual test cases in Julia, so I'd likely set it to always be identical to `message`. I'm not sure how other languages handle it, perhaps a default fallback would make sense.

Reply:
@SaschaMann Could you paste a screenshot of the test output in Julia so I can get a feel for how it's presented pls?

Reply:
First a toy example on how the tests are defined because that explains why they don't have a name:

```julia
@testset "Maths" begin
    @test 1 + 1 == 2
    @test 1 - 1 == 0
end
```

If everything passes:

```
$ julia runtests.jl
Test Summary:        | Pass  Total
One- & Zero-elements |    2      2
Test Summary: | Pass  Total
Arithmetic    |   27     27
Test Summary:  | Pass  Total
Absolute value |    4      4
Test Summary:             | Pass  Total
Reduction to lowest terms |   14     14
Test Summary: |  Pass  Total
Ordering      | 94320  94320
Test Summary:           | Pass  Total
Showing RationalNumbers |    2      2
```

By default it only shows the top level of testsets (tests that are grouped together). If a test in an inner set fails, it shows more levels. Here a test case in the `Arithmetic` -> `Addition` set fails:

```
Test Summary:        | Pass  Total
One- & Zero-elements |    2      2
Addition: Test Failed at h:\Sascha\Documents\Julia\Exercism_xjulia\exercises\rational-numbers\runtests.jl:15
  Expression: RationalNumber(1, 2) + RationalNumber(2, 3) == RationalNumber(5, 6)
   Evaluated: 7//6 == 5//6
Stacktrace:
 [1] top-level scope at h:\Sascha\Documents\Julia\Exercism_xjulia\exercises\rational-numbers\runtests.jl:15
 [2] top-level scope at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\Test\src\Test.jl:1083
 [3] top-level scope at h:\Sascha\Documents\Julia\Exercism_xjulia\exercises\rational-numbers\runtests.jl:15
 [4] top-level scope at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\Test\src\Test.jl:1083
 [5] top-level scope at h:\Sascha\Documents\Julia\Exercism_xjulia\exercises\rational-numbers\runtests.jl:14
Test Summary:    | Pass  Fail  Total
Arithmetic       |   26     1     27
  Addition       |    3     1      4
  Subtraction    |    4            4
  Multiplication |    6            6
  Exponentiation |    9            9
  Division       |    4            4
ERROR: LoadError: Some tests did not pass: 26 passed, 1 failed, 0 errored, 0 broken.
in expression starting at h:\Sascha\Documents\Julia\Exercism_xjulia\exercises\rational-numbers\runtests.jl:13
```

If a non-grouped test fails:
```
Test Failed at h:\Sascha\Documents\Julia\Exercism_xjulia\exercises\rational-numbers\runtests.jl:6
  Expression: 1 == 2
   Evaluated: 1 == 2
ERROR: LoadError: There was an error during testing
in expression starting at h:\Sascha\Documents\Julia\Exercism_xjulia\exercises\rational-numbers\runtests.jl:6
```

All tests in a set are evaluated but as soon as a set has failures, it stops running the script.

You can find the test file used for the examples above here: https://github.com/exercism/julia/blob/master/exercises/rational-numbers/runtests.jl

---

It would be possible to require every testcase to be part of a testset and then use the name of the testset to make it fit this format. However, this doesn't match how people write test suites for Julia in the "real world" and would still require flattening it which will reduce readability.

Reply:
Here's an example on how some CI runners display output based on a JUnit file. This isn't commonly used but it's quite similar to our use case:

https://github.com/oxinabox/TestReports.jl#example-of-use

Reply:
Interesting feature! If I look at the test output example, they are only showing the name of the failed tests. I saw that the line number is included in the test failure. Maybe it would be useful if the line number is included in the JSON output? Adding the line number would also allow for somewhat fancy integration with the website, where we can display the status of a test very precisely.

Reply:
In Go I can get the names only from the test output if we rewrite all the tests to make each test a subtest:

```go
func TestMain_Foo(t *testing.T) {
	for _, tt := range tests1 {
		assert.False(t, tt.fail)
	}
}
```

needs to be

```go
func TestMain_Foo(t *testing.T) {
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.False(t, tt.fail)
		})
	}
}
```

Is this something we can do @iHiD ?

Also it might be good to make the name optional for the start so we can change the tests step by step.

Reply:
This works well for me in go without requiring any code changes:

```sh
$ go test . -v | grep -E (RUN|PASS|FAIL)
```